buildflow/core/app/runtime/actors/runtime.py

```diff
@@ -159,55 +163,71 @@ def is_active(self):
 
     async def _runtime_autoscale_loop(self):
         logging.info("Runtime checkin loop started...")
-        while self._status == RuntimeStatus.RUNNING:
-            for processor_pool in self._processor_pool_refs:
-                if self._status != RuntimeStatus.RUNNING:
-                    break
-
-                try:
-                    processor_snapshot: ProcessorSnapshot = (
-                        await processor_pool.actor_handle.snapshot.remote()
-                    )
-                except RayActorError:
-                    logging.exception("process actor unexpectedly died. will restart.")
-                    options = self.options.processor_options[
-                        processor_pool.processor.processor_id
-                    ]
-                    processor_pool.actor_handle = (
-                        PipelineProcessorReplicaPoolActor.remote(
-                            self.run_id,
-                            processor_pool.processor,
-                            options,
+        last_autoscale_event = datetime.utcnow()
+        # We keep running the loop while the job is running or draining to ensure
+        # we don't exit the main process before the drain is complete.
+        while (
+            self._status == RuntimeStatus.RUNNING
+            or self._status == RuntimeStatus.DRAINING
+        ):
+            # Only run the autoscale loop when the runtime is running, this prevents
+            # us from scaling while we are draining.
+            if (
+                self._status == RuntimeStatus.RUNNING
+                and last_autoscale_event + self._autoscale_frequency > datetime.utcnow()
+            ):
+                for processor_pool in self._processor_pool_refs:
+                    if self._status != RuntimeStatus.RUNNING:
+                        break
+
+                    try:
+                        processor_snapshot: ProcessorSnapshot = (
+                            await processor_pool.actor_handle.snapshot.remote()
                         )
+                    except RayActorError:
```

Comment:
nit: Is there a way to check if this is a dead actor error specifically? Wondering if this might include other errors and then we'll have a runaway actor (or we could do ray.kill to make sure its dead

Reply:
According to the docs this is the error that gets thrown communicating with a dead actor: https://docs.ray.io/en/latest/ray-core/api/doc/ray.exceptions.RayActorError.html

We could potentially use the state API and check if it's alive but I believe that requires a network hop so probably not worth the added cost since this is a rare case.